### PR TITLE
Return (non-)fungible resource balances from all Vaults under requested Component

### PR DESCRIPTION
**Pull request by KrzysztofLabus-RDX**

### CODE REVIEW

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/PostgresLedgerExtenderService.cs [hunk 1]
```diff
@@ -858,6 +958,24 @@ void AggregateEntityResourceInternal(long entityId, long resourceEntityId)
                 })
                 .ToList();
 
+            var entityNonFungibleVaultHistoryToAdd = nonFungibleVaultChanges
+                .Select(e =>
+                {
+                    AggregateEntityResource(e.ReferencedVault, e.ReferencedResource, e.StateVersion, x => x.NonFungibleResourceEntityIds, x => x.NonFungibleResourceLastUpdateStateVersions);
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/PostgresLedgerExtenderService.cs:964]: It took my a lot of time to understand what's going on in here. 

As you already mentioned `PostgresLedgerExtenderService` deserves some care/refactoring.  Just wanted to raise it here so we can schedule/agree on some actions to perform.

We're calling local function which is placed in the middle of upper function (in one of `case` scenarios), which then calls yet another inner local function which manipulates some local variable from 3 scopes up. Overall complexity of entire logic here is very hard to understand and I bet it's gonna be almost impossible to write/change any code here for me.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/WriteHelper.cs [hunk 2]
```diff
@@ -397,25 +397,51 @@ public async Task<int> CopyEntityResourceAggregateHistory(ICollection<EntityReso
...
+            await writer.WriteAsync(e.NonFungibleResourceLastUpdateStateVersions.ToArray(), NpgsqlDbType.Array | NpgsqlDbType.Bigint, token);
         }
 
         await writer.CompleteAsync(token);
 
         return entities.Count;
     }
 
-    public async Task<int> CopyEntityResourceHistory(ICollection<EntityFungibleResourceHistory> fungibleEntities, ICollection<EntityNonFungibleResourceHistory> nonFungibleEntities, CancellationToken token)
+    public async Task<int> CopyEntityResourceVaultAggregateHistory(ICollection<EntityResourceVaultAggregateHistory> entities, CancellationToken token)
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/WriteHelper.cs:410]: I'm not sure we should call these methods `Copy`. What these methods do is write/save data into db. It's internal that it performs postgres copy operation.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs [hunk 3]
```diff
@@ -641,15 +669,17 @@ LIMIT 1
     private async Task<GatewayModel.NonFungibleIdsCollection> GetNonFungibleIdsSlice(long entityId, long resourceEntityId, int offset, int limit,
         GatewayModel.LedgerState ledgerState, CancellationToken token)
     {
+        // TODO this query does not account for multiple Vaults that could store NFIDs
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:672]: if you can add ticket number as we agreed for that todo.

**Review by PawelPawelec-RDX — COMMENTED** [review]
Just minor fixes to do, most of my comments were made to open discussion, my intention is not to fix them all withing this PR.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/ReadHelper.cs [hunk 4]
```diff
@@ -123,6 +123,49 @@ LIMIT 1
...
+            data.Add(new EntityResourceVaultLookup(change.ReferencedVault.DatabaseGlobalAncestorId, change.ReferencedResource.DatabaseId));
+        }
+
+        foreach (var change in nonFungibleVaultChanges)
+        {
+            data.Add(new EntityResourceVaultLookup(change.ReferencedVault.DatabaseOwnerAncestorId, change.ReferencedResource.DatabaseId));
+            data.Add(new EntityResourceVaultLookup(change.ReferencedVault.DatabaseGlobalAncestorId, change.ReferencedResource.DatabaseId));
+        }
+
+        var eid = new List<long>();
```
[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/ReadHelper.cs:142]: Oh my... this was literally me being bored with copyin'n'pasting code just to get things done the ugly way. Those must be properly named. Short names are suitable for super obvious cases or 5-10 LoC code blocks.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/WriteHelper.cs (hunk 2 above)
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/WriteHelper.cs:410]: Going to change that as (part of) separate PR to ease up further merger to `release/betanet-v2`

**Review by PawelPawelec-RDX — APPROVED** [review]

**Pull request merged**